File: python/heap.py
"""
Max Heap Implementation in Python
Based on Java MaxHeap implementation
"""

from typing import TypeVar, Generic, List, Optional
from abc import ABC, abstractmethod

T = TypeVar('T')


class MaxHeap(Generic[T]):
    """Max Heap implementation using dynamic array"""

    def __init__(self, capacity: int = 10):
        self._data: List[Optional[T]] = [None] * capacity
        self._size: int = 0

    def size(self) -> int:
        """Get heap size"""
        return self._size

    def is_empty(self) -> bool:
        """Check if heap is empty"""
        return self._size == 0

    def add(self, e: T) -> None:
        """Add element to heap"""
        if self._size == len(self._data):
            self._resize(2 * len(self._data))
        self._data[self._size] = e
        self._sift_up(self._size)
        self._size += 1

    def max(self) -> T:
        """Get maximum element (root)"""
        if self._size == 0:
            raise IndexError("Heap is empty")
        return self._data[0]  # type: ignore

    def extract_max(self) -> T:
        """Extract and return maximum element"""
        if self._size == 0:
            raise IndexError("Heap is empty")

        ret = self._data[0]
        self._swap(0, self._size - 1)
        self._data[self._size - 1] = None
        self._size -= 1
        self._sift_down(0)

        if self._size <= len(self._data) // 4 and len(self._data) // 2 != 0:
            self._resize(len(self._data) // 2)

        return ret  # type: ignore

    def replace(self, e: T) -> T:
        """Replace root element with new element and return old root"""
        if self._size == 0:
            raise IndexError("Heap is empty")
        ret = self._data[0]
        self._data[0] = e
        self._sift_down(0)
        return ret  # type: ignore

    def _sift_up(self, index: int) -> None:
        """Sift element up to maintain heap property"""
        while index > 0 and self._data[self._parent(index)] < self._data[index]:  # type: ignore
            self._swap(self._parent(index), index)
            index = self._parent(index)

    def _sift_down(self, index: int) -> None:
        """Sift element down to maintain heap property"""
        while self._left(index) < self._size:
            j = self._left(index)
            # Find the larger child
            if j + 1 < self._size and self._data[j + 1] > self._data[j]:  # type: ignore
                j = self._right(index)

            # If parent is larger or equal, stop
            if self._data[index] >= self._data[j]:  # type: ignore
                break

            self._swap(index, j)
            index = j

    def _parent(self, index: int) -> int:
        """Get parent index"""
        if index <= 0:
            raise ValueError("Index has no parent")
        return (index - 1) // 2

    def _left(self, index: int) -> int:
        """Get left child index"""
        return index * 2 + 1

    def _right(self, index: int) -> int:
        """Get right child index"""
        return index * 2 + 2

    def _swap(self, i: int, j: int) -> None:
        """Swap two elements"""
        self._data[i], self._data[j] = self._data[j], self._data[i]

    def _resize(self, new_capacity: int) -> None:
        """Resize internal array"""
        new_data: List[Optional[T]] = [None] * new_capacity
        for i in range(self._size):
            new_data[i] = self._data[i]
        self._data = new_data

    def to_list(self) -> List[T]:
        """Convert heap to list"""
        return [self._data[i] for i in range(self._size)]  # type: ignore

    def __str__(self) -> str:
        return f"MaxHeap({self.to_list()})"

    def __repr__(self) -> str:
        return self.__str__()

    def __len__(self) -> int:
        return self._size

    def __bool__(self) -> bool:
        return not self.is_empty()


# Priority Queue based on Max Heap
class PriorityQueue(Generic[T]):
    """Priority Queue implementation using Max Heap"""

    def __init__(self, capacity: int = 10):
        self._heap: MaxHeap[T] = MaxHeap[T](capacity)

    def enqueue(self, e: T) -> None:
        """Add element to queue"""
        self._heap.add(e)

    def dequeue(self) -> T:
        """Remove and return highest priority element"""
        return self._heap.extract_max()

    def front(self) -> T:
        """Get highest priority element without removing"""
        return self._heap.max()

    def size(self) -> int:
        """Get queue size"""
        return self._heap.size()

    def is_empty(self) -> bool:
        """Check if queue is empty"""
        return self._heap.is_empty()

    def __str__(self) -> str:
        return f"PriorityQueue({self._heap.to_list()})"


if __name__ == "__main__":
    # Test MaxHeap
    print("Testing MaxHeap:")
    heap = MaxHeap[int]()
    for i in range(10):
        heap.add(i)
    print(f"Heap: {heap}")
    print(f"Max: {heap.max()}")

    for _ in range(5):
        print(f"Extract max: {heap.extract_max()}")
    print(f"After 5 extracts: {heap}")

    heap.replace(100)
    print(f"After replace(100): {heap}")

    # Test PriorityQueue
    print("\nTesting PriorityQueue:")
    pq = PriorityQueue[int]()
    pq.enqueue(3)
    pq.enqueue(1)
    pq.enqueue(4)
    pq.enqueue(2)
    print(f"PriorityQueue: {pq}")
    print(f"Front: {pq.front()}")

    while not pq.is_empty():
        print(f"Dequeue: {pq.dequeue()}")

File: golang/Heap/heap.go
package main

import (
	"fmt"
)

// MaxHeap 最大堆实现
type MaxHeap[E any] struct {
	data      []E
	size      int
	lessFunc  func(E, E) bool
}

// NewMaxHeap 创建一个新的最大堆
// lessFunc 用于比较两个元素，返回 true 表示 a < b
func NewMaxHeap[E any](lessFunc func(E, E) bool, capacity int) *MaxHeap[E] {
	return &MaxHeap[E]{
		data:     make([]E, capacity),
		size:     0,
		lessFunc: lessFunc,
	}
}

// NewMaxHeapDefault 创建默认容量的最大堆
func NewMaxHeapDefault[E any](lessFunc func(E, E) bool) *MaxHeap[E] {
	return NewMaxHeap[E](lessFunc, 10)
}

func (h *MaxHeap[E]) Size() int {
	return h.size
}

func (h *MaxHeap[E]) IsEmpty() bool {
	return h.size == 0
}

func (h *MaxHeap[E]) Add(e E) {
	if h.size == len(h.data) {
		h.resize(2 * len(h.data))
	}
	h.data[h.size] = e
	h.siftUp(h.size)
	h.size++
}

func (h *MaxHeap[E]) Max() E {
	if h.IsEmpty() {
		panic("Heap is empty")
	}
	return h.data[0]
}

func (h *MaxHeap[E]) ExtractMax() E {
	if h.IsEmpty() {
		panic("Heap is empty")
	}

	ret := h.data[0]
	h.swap(0, h.size-1)
	h.size--
	h.siftDown(0)

	if h.size <= len(h.data)/4 && len(h.data)/2 != 0 {
		h.resize(len(h.data) / 2)
	}

	return ret
}

func (h *MaxHeap[E]) Replace(e E) E {
	if h.IsEmpty() {
		panic("Heap is empty")
	}

	ret := h.data[0]
	h.data[0] = e
	h.siftDown(0)
	return ret
}

// siftUp 元素上浮
func (h *MaxHeap[E]) siftUp(index int) {
	for index > 0 && h.lessFunc(h.data[h.parent(index)], h.data[index]) {
		h.swap(h.parent(index), index)
		index = h.parent(index)
	}
}

// siftDown 元素下沉
func (h *MaxHeap[E]) siftDown(index int) {
	for h.left(index) < h.size {
		j := h.left(index)
		// 找到左右子节点中较大的那个
		if j+1 < h.size && h.lessFunc(h.data[j], h.data[j+1]) {
			j = h.right(index)
		}

		// 如果父节点已经大于等于较大的子节点，停止
		if h.lessFunc(h.data[j], h.data[index]) {
			break
		}

		h.swap(index, j)
		index = j
	}
}

func (h *MaxHeap[E]) parent(index int) int {
	if index <= 0 {
		panic("Index has no parent")
	}
	return (index - 1) / 2
}

func (h *MaxHeap[E]) left(index int) int {
	return index*2 + 1
}

func (h *MaxHeap[E]) right(index int) int {
	return index*2 + 2
}

func (h *MaxHeap[E]) swap(i, j int) {
	h.data[i], h.data[j] = h.data[j], h.data[i]
}

func (h *MaxHeap[E]) resize(newCapacity int) {
	newData := make([]E, newCapacity)
	for i := 0; i < h.size; i++ {
		newData[i] = h.data[i]
	}
	h.data = newData
}

func (h *MaxHeap[E]) ToSlice() []E {
	result := make([]E, h.size)
	for i := 0; i < h.size; i++ {
		result[i] = h.data[i]
	}
	return result
}

func (h *MaxHeap[E]) String() string {
	return fmt.Sprintf("MaxHeap%v", h.ToSlice())
}

// PriorityQueue 基于最大堆的优先队列
type PriorityQueue[E any] struct {
	heap *MaxHeap[E]
}

// NewPriorityQueue 创建一个新的优先队列
func NewPriorityQueue[E any](lessFunc func(E, E) bool, capacity int) *PriorityQueue[E] {
	return &PriorityQueue[E]{
		heap: NewMaxHeap[E](lessFunc, capacity),
	}
}

// NewPriorityQueueDefault 创建默认容量的优先队列
func NewPriorityQueueDefault[E any](lessFunc func(E, E) bool) *PriorityQueue[E] {
	return NewPriorityQueue[E](lessFunc, 10)
}

func (pq *PriorityQueue[E]) Enqueue(e E) {
	pq.heap.Add(e)
}

func (pq *PriorityQueue[E]) Dequeue() E {
	return pq.heap.ExtractMax()
}

func (pq *PriorityQueue[E]) Front() E {
	return pq.heap.Max()
}

func (pq *PriorityQueue[E]) Size() int {
	return pq.heap.Size()
}

func (pq *PriorityQueue[E]) IsEmpty() bool {
	return pq.heap.IsEmpty()
}

func (pq *PriorityQueue[E]) String() string {
	return fmt.Sprintf("PriorityQueue%v", pq.heap.ToSlice())
}

func main() {
	fmt.Println("=== Go语言堆和优先队列示例 ===")

	// 创建用于整数的比较函数
	intLess := func(a, b int) bool {
		return a < b
	}

	// 测试最大堆
	fmt.Println("\n--- 最大堆 ---")
	heap := NewMaxHeapDefault[intLess](intLess)
	for i := 0; i < 10; i++ {
		heap.Add(i)
	}
	fmt.Printf("堆: %s\n", heap.String())
	fmt.Printf("最大值: %d\n", heap.Max())

	fmt.Println("\n取出5个最大值:")
	for i := 0; i < 5; i++ {
		fmt.Printf("ExtractMax: %d\n", heap.ExtractMax())
	}
	fmt.Printf("剩余: %s\n", heap.String())

	// 测试优先队列
	fmt.Println("\n--- 优先队列 ---")
	pq := NewPriorityQueueDefault[intLess](intLess)
	pq.Enqueue(3)
	pq.Enqueue(1)
	pq.Enqueue(4)
	pq.Enqueue(2)
	fmt.Printf("优先队列: %s\n", pq.String())

	fmt.Println("按优先级出队:")
	for !pq.IsEmpty() {
		fmt.Printf("Dequeue: %d\n", pq.Dequeue())
	}

	fmt.Println("\n=== 示例程序结束 ===")
}
